### PR TITLE
feat: Granular permissions 

### PR DESCRIPTION
**Pull request by JanCizmar**
The licence modification in root is made only for purposes of this PR, the licence will move to ee spcecific folder later.

### CODE REVIEW

File: backend/data/src/main/kotlin/io/tolgee/model/Permission.kt [hunk 1]
```diff
@@ -22,67 +46,122 @@ class Permission(
...
+   * When specified, user is restricted to edit specific language translations.
    */
   @ManyToMany(fetch = FetchType.EAGER)
-  var languages: MutableSet<Language> = mutableSetOf()
+  @JoinTable(name = "permission_languages", inverseJoinColumns = [JoinColumn(name = "languages_id")])
+  var translateLanguages: MutableSet<Language> = mutableSetOf()
+
+  /**
+   * Languages for TRANSLATIONS_EDIT scope.
+   * When specified, user is restricted to edit specific language translations.
```
Comment (cyyynthia) [inline backend/data/src/main/kotlin/io/tolgee/model/Permission.kt:88]:
```suggestion
   * Languages for TRANSLATIONS_VIEW scope.
   * When specified, user is restricted to access specific language translations.
```

File: backend/data/src/main/kotlin/io/tolgee/model/Permission.kt [hunk 2]
```diff
@@ -22,67 +46,122 @@ class Permission(
...
+   */
+  @ManyToMany(fetch = FetchType.EAGER)
+  var viewLanguages: MutableSet<Language> = mutableSetOf()
+
+  /**
+   * Languages for TRANSLATIONS_EDIT scope.
+   * When specified, user is restricted to edit specific language translations.
```
Comment (cyyynthia) [inline backend/data/src/main/kotlin/io/tolgee/model/Permission.kt:95]:
```suggestion
   * Languages for TRANSLATIONS_STATE_EDIT scope.
   * When specified, user is restricted to edit specific language translation states.
```

File: webapp/package.json [hunk 3]
```diff
@@ -65,10 +68,12 @@
     "test": "craco test",
     "schema": "openapi-typescript http://localhost:8080/v3/api-docs/All%20Internal%20-%20for%20Tolgee%20Web%20application --output ./src/service/apiSchema.generated.ts",
     "billing-schema": "openapi-typescript http://localhost:8080/v3/api-docs/V2%20Billing --output ./src/service/billingApiSchema.generated.ts",
+    "scopes": "",
```
Comment (cyyynthia) [inline webapp/package.json:71]: Looks like a leftover that should be removed

File: webapp/scripts/generateScopes.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+const fetch = (...args) =>
+  import('node-fetch').then(({ default: fetch }) => fetch(...args));
+
+const host = process.env.HOST ?? 'http://localhost:8080';
+
+(async () => {
+  const scopes = await getScopes();
+  const content = JSON.stringify(scopes, null, 2);
+  fs.writeFileSync(
+    path.resolve(__dirname + '/../src/constants/scopes.generated.json'),
```
Comment (cyyynthia) [inline webapp/scripts/generateScopes.js:12]:
```suggestion
    path.resolve(__dirname, '../src/constants/scopes.generated.json'),
```

File: webapp/src/constants/scopeCategories.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import { expandScope, Scope } from 'tg.component/permissions/permissionHelper';
+
+export function getScopeCategories(t: (key: string) => string) {
+  const scopeCategories = [
+    {
+      name: t('scope_category_project_management'),
```
Comment (cyyynthia) [inline webapp/src/constants/scopeCategories.ts:6]: :eyes:
```suggestion
      // @tolgee-key scope_category_project_management
      name: t('scope_category_project_management'),
```

File: webapp/src/views/projects/translations/TranslationHeader/TranslationControls.tsx [hunk 6]
```diff
@@ -13,6 +13,7 @@ import {
 } from '../context/TranslationsContext';
 import { Filters } from '../Filters/Filters';
 import { StickyHeader } from './StickyHeader';
+import { StickyHeader } from './StickyHeader';
```
Comment (cyyynthia) [inline webapp/src/views/projects/translations/TranslationHeader/TranslationControls.tsx:16]: Duplicated import

File: backend/data/src/main/kotlin/io/tolgee/model/enums/Scope.kt [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import io.tolgee.constants.Message
+import io.tolgee.exceptions.BadRequestException
+import io.tolgee.exceptions.NotFoundException
+
+enum class Scope(
+  @get:JsonValue
+  var value: String
+) {
+  // !!!!
+  // Don't change the order of the enum, it's stored as ORDINAL is ApiKey entity
```
Comment (cyyynthia) [inline backend/data/src/main/kotlin/io/tolgee/model/enums/Scope.kt:13]: This sound unnecessarily fragile.. Wouldn't it be possible to store the value in db?

**Review by cyyynthia — CHANGES_REQUESTED** [review]

File: backend/app/src/main/kotlin/io/tolgee/api/v2/controllers/AutoTranslationController.kt [hunk 8]
```diff
@@ -39,8 +38,8 @@ class AutoTranslationController(
   private val securityService: SecurityService
 ) {
   @PutMapping("")
-  @AccessWithProjectPermission(Permission.ProjectPermissionType.TRANSLATE)
-  @AccessWithApiKey(scopes = [ApiScope.TRANSLATIONS_EDIT])
+  @AccessWithProjectPermission(Scope.TRANSLATIONS_EDIT)
+  @AccessWithApiKey(scopes = [Scope.TRANSLATIONS_EDIT])
```
[the remark this answers is not in the document]
Comment (JanCizmar) [reply backend/app/src/main/kotlin/io/tolgee/api/v2/controllers/AutoTranslationController.kt:42]: Yes.

File: backend/app/src/main/kotlin/io/tolgee/api/v2/hateoas/permission/IPermissionModel.kt [hunk 9]
```diff
@@ -0,0 +1,51 @@
+package io.tolgee.api.v2.hateoas.permission
+
+import io.swagger.v3.oas.annotations.media.Schema
+import io.tolgee.model.enums.ProjectPermissionType
+import io.tolgee.model.enums.Scope
+
+interface IPermissionModel {
```
[the remark this answers is not in the document]
Comment (JanCizmar) [reply backend/app/src/main/kotlin/io/tolgee/api/v2/hateoas/permission/IPermissionModel.kt:7]: Well, I've deleted the granular field. But I am not sure what's unclear in the descriptions. Can you help me there?

File: webapp/src/constants/scopeCategories.ts (hunk 5 above)
Comment (JanCizmar) [reply webapp/src/constants/scopeCategories.ts:6]: Why?

Comment (cyyynthia) [reply webapp/src/constants/scopeCategories.ts:6]: The `t` doesn't come from a direct import, so the CLI would be unable to detect these. Adding annotations would make sure the CLI sees these and extracts them

Comment (JanCizmar) [reply webapp/src/constants/scopeCategories.ts:6]: Oh. Right!

**Pull request closed without merging**